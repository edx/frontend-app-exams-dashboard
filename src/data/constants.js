export const RequestStates = {
  inactive: 'inactive',
  pending: 'pending',
  completed: 'completed',
  failed: 'failed',
};

export const RequestKeys = {
  fetchAllowances: 'fetchAllowances',
  fetchCourseExams: 'fetchCourseExams',
  fetchExamAttempts: 'fetchExamAttempts',
  deleteAllowance: 'deleteAllowance',
  deleteExamAttempt: 'deleteExamAttempt',
  modifyExamAttempt: 'modifyExamAttempt',
  createAllowance: 'createAllowance',
};

export const ExamAttemptActions = {
  // NOTE: These are the "staff-only" actions, which are meant to be performed from this dashboard.
  verify: 'verify',
  reject: 'reject',
};

export const ExamAttemptStatus = {
  created: 'created',
  download_software_clicked: 'download_software_clicked',
  ready_to_start: 'ready_to_start',
  started: 'started',
  ready_to_submit: 'ready_to_submit',
  timed_out: 'timed_out',
  submitted: 'submitted',
  verified: 'verified',
  rejected: 'rejected',
  expired: 'expired',
  second_review_required: 'second_review_required',
  error: 'error',
};

export const ErrorStatuses = {
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
};
