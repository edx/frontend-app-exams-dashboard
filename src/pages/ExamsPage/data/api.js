import { ensureConfig, getConfig } from '@edx/frontend-platform';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

export function getExamsBaseUrl() {
  ensureConfig([
    'EXAMS_BASE_URL',
  ]);
  return getConfig().EXAMS_BASE_URL;
}

export async function getCourseExams(courseId) {
  const url = `${getExamsBaseUrl()}/api/v1/exams/course_id/${courseId}/`;
  const response = await getAuthenticatedHttpClient().get(url);
  return response.data;
}

export async function getExamAttempts(courseId, examId) {
  const url = `${getExamsBaseUrl()}/api/v1/instructor_view/course_id/${courseId}/attempts?exam_id=${examId}&limit=1000000`;
  const response = await getAuthenticatedHttpClient().get(url);
  return response.data;
}

export async function deleteExamAttempt(attemptId) {
  const url = `${getExamsBaseUrl()}/api/v1/exams/attempt/${attemptId}`;
  const response = await getAuthenticatedHttpClient().delete(url);
  return response.data;
}

export async function modifyExamAttempt(attemptId, action) {
  const url = `${getExamsBaseUrl()}/api/v1/exams/attempt/${attemptId}`;
  const payload = { action };
  const response = await getAuthenticatedHttpClient().put(url, payload);
  return response.data;
}
