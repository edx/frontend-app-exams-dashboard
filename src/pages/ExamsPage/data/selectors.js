export const courseExamsList = state => state.exams.examsList;

// derived from currentExamIndex rather than duplicating in state
export const currentExam = state => state.exams.examsList[state.exams.currentExamIndex];

export const courseExamAttemptsList = state => state.exams.attemptsList;

export const courseId = state => state.exams.courseId;

export const courseAllowancesList = state => state.exams.allowancesList;
