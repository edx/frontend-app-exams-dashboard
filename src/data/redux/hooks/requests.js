import { useSelector, useDispatch } from 'react-redux';

import * as redux from 'data/redux';
import * as module from './requests'; // eslint-disable-line import/no-self-import

const selectors = redux.selectors.requests;
const actions = redux.actions.requests;

export const statusSelector = selector => (requestName) => useSelector(selector(requestName));
export const useRequestIsPending = module.statusSelector(selectors.isPending);
export const useRequestIsFailed = module.statusSelector(selectors.isFailed);
export const useRequestIsCompleted = module.statusSelector(selectors.isCompleted);
export const useRequestIsInactive = module.statusSelector(selectors.isInactive);
export const useRequestError = module.statusSelector(selectors.error);
export const useRequestErrorCode = module.statusSelector(selectors.errorCode);
export const useRequestErrorStatus = module.statusSelector(selectors.errorStatus);
export const useRequestData = module.statusSelector(selectors.data);

export const useMakeNetworkRequest = () => {
  const dispatch = useDispatch();
  return ({
    requestKey,
    promise,
    onSuccess,
    onFailure,
  }) => {
    dispatch(actions.startRequest({ requestKey }));
    return promise.then((response) => {
      if (onSuccess) { onSuccess(response); }
      dispatch(actions.completeRequest({ requestKey, response }));
    }).catch((error) => {
      // Parse backend error from axios object
      let errorParsed = error;
      if (error?.customAttributes?.httpErrorResponseData) {
        errorParsed = JSON.parse(error.customAttributes.httpErrorResponseData);
      }
      if (onFailure) { onFailure(errorParsed); }
      dispatch(actions.failRequest({ requestKey, error: errorParsed }));
    });
  };
};

export const useClearRequest = (requestKey) => {
  const dispatch = useDispatch();
  return () => {
    dispatch(actions.clearRequest({ requestKey }));
  };
};
