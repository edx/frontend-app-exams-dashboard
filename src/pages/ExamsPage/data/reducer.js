import { createSlice } from '@reduxjs/toolkit';
import * as constants from 'data/constants';

export const initialState = {
  courseId: null,
  currentExamIndex: null,
  examsList: [],
  attemptsList: [],
};

const getStatusFromAction = (action, status) => {
  switch (action) {
    case constants.ExamAttemptActions.verify:
      return constants.ExamAttemptStatus.verified;
    case constants.ExamAttemptActions.reject:
      return constants.ExamAttemptStatus.rejected;
    default:
      // If invalid action, return the same status as before.
      return status;
  }
};

const getCurrentExamIndex = (examsList, examId) => {
  const index = examsList.findIndex(exam => exam.id === examId.payload);
  if (index > -1) {
    return index;
  }
  return null;
};

const slice = createSlice({
  name: 'exams',
  initialState,
  reducers: {
    setCourseId: (state, courseId) => ({
      ...state,
      courseId: courseId.payload,
    }),
    loadExams: (state, { payload }) => ({
      ...state,
      examsList: payload.map((exam) => ({
        id: exam.id,
        name: exam.exam_name,
      })),
    }),
    loadExamAttempts: (state, { payload }) => ({
      ...state,
      attemptsList: payload?.results.map((attempt) => {
        const data = {
          attempt_id: attempt.attempt_id,
          exam_name: attempt.exam_display_name,
          username: attempt.username,
          time_limit: attempt.allowed_time_limit_mins,
          exam_type: attempt.exam_type,
          started_at: attempt.start_time,
          completed_at: attempt.end_time,
          status: attempt.attempt_status,
        };
        // Only add ACS review values if they exist
        if (attempt.proctored_review) {
          Object.assign(
            data,
            { severity: attempt.proctored_review.severity },
            { submission_reason: attempt.proctored_review.submission_reason },
          );
        }
        return data;
      }),
    }),
    deleteExamAttempt: (state, attemptId) => ({
      ...state,
      attemptsList: state.attemptsList.filter(attempt => attempt.attempt_id !== attemptId.payload),
    }),
    modifyExamAttemptStatus: (state, { payload }) => ({
      ...state,
      attemptsList: state.attemptsList.map((attempt) => {
        // Set the status of the modified attempt to verified or rejected
        if (attempt.attempt_id === payload.attemptId) {
          return {
            ...attempt,
            status: getStatusFromAction(payload.action, attempt.status),
          };
        }
        // Keep all other attempts as is
        return attempt;
      }),
    }),
    setCurrentExam: (state, examId) => ({
      ...state,
      currentExamIndex: getCurrentExamIndex(state.examsList, examId),
    }),
  },
});

export const {
  loadExams,
  loadExamAttempts,
  deleteExamAttempt,
  modifyExamAttemptStatus,
  setCurrentExam,
  setCourseId,
} = slice.actions;

export const {
  reducer,
} = slice;
