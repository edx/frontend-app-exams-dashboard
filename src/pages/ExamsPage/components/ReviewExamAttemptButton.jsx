import PropTypes from 'prop-types';

import {
  Button, useToggle, ModalDialog, ActionRow,
} from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';
import { Warning } from '@edx/paragon/icons';
import * as constants from 'data/constants';
import { useModifyExamAttempt } from '../hooks';
import messages from '../messages';

const ReviewExamAttemptButton = ({
  username, examName, attemptId, severity, submissionReason,
}) => {
  const [isOpen, open, close] = useToggle(false);
  const modifyExamAttempt = useModifyExamAttempt();
  const { formatMessage } = useIntl();

  return (
    <>
      <div className="d-flex text-danger">
        <Button variant="link" size="sm" className="text-danger" onClick={open}>
          <Warning />
          {formatMessage(messages.ReviewExamAttemptButtonTitle)}
        </Button>
      </div>
      <ModalDialog
        title="my dialog"
        isOpen={isOpen}
        onClose={close}
        size="md"
        variant="default"
        hasCloseButton
        isFullscreenOnMobile
      >
        <ModalDialog.Header>
          <ModalDialog.Title>
            {formatMessage(messages.ReviewExamAttemptButtonModalTitle)}
          </ModalDialog.Title>
        </ModalDialog.Header>

        <ModalDialog.Body>
          <p>{formatMessage(messages.ReviewExamAttemptButtonModalBody)}</p>
          <ul>
            <li>{formatMessage(messages.Username)}{username}</li>
            <li>{formatMessage(messages.ExamName)}{examName}</li>
            <li>{formatMessage(messages.SuspicionLevel)}{severity}</li>
            <li>{formatMessage(messages.SubmissionReason)}{submissionReason}</li>
          </ul>
        </ModalDialog.Body>

        <ModalDialog.Footer>
          <ActionRow>
            <ModalDialog.CloseButton variant="tertiary">
              {formatMessage(messages.ReviewExamAttemptButtonCancel)}
            </ModalDialog.CloseButton>
            <Button
              variant="primary"
              onClick={e => { // eslint-disable-line no-unused-vars
                modifyExamAttempt(attemptId, constants.ExamAttemptActions.verify);
              }}
            >
              {formatMessage(messages.ReviewExamAttemptButtonVerify)}
            </Button>
            <Button
              variant="primary"
              onClick={e => { // eslint-disable-line no-unused-vars
                modifyExamAttempt(attemptId, constants.ExamAttemptActions.reject);
              }}
            >
              {formatMessage(messages.ReviewExamAttemptButtonReject)}
            </Button>
          </ActionRow>
        </ModalDialog.Footer>
      </ModalDialog>
    </>
  );
};

ReviewExamAttemptButton.propTypes = {
  username: PropTypes.string.isRequired,
  examName: PropTypes.string.isRequired,
  attemptId: PropTypes.number.isRequired,
  severity: PropTypes.number.isRequired,
  submissionReason: PropTypes.string.isRequired,
};

export default ReviewExamAttemptButton;
