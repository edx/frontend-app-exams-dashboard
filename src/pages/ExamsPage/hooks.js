import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RequestKeys } from 'data/constants';

import * as reduxHooks from 'data/redux/hooks';

import * as api from './data/api';
import * as selectors from './data/selectors';
import * as reducer from './data/reducer';

import * as module from './hooks'; // eslint-disable-line import/no-self-import

export const state = {
  // example of component state that does not need to be in redux
  exampleValue: (val) => React.useState(val), // eslint-disable-line
};

export const useFetchCourseExams = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (courseId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchCourseExams,
      promise: api.getCourseExams(courseId),
      onSuccess: (response) => dispatch(reducer.loadExams(response)),
    })
  );
};

export const useFetchExamAttempts = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (examId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchExamAttempts,
      promise: api.getExamAttempts(examId),
      onSuccess: (response) => dispatch(reducer.loadExamAttempts(response)),
    })
  );
};

export const useInitializeExamsPage = (courseId) => {
  const fetchCourseExams = module.useFetchCourseExams();
  React.useEffect(() => { fetchCourseExams(courseId); }, []); // eslint-disable-line react-hooks/exhaustive-deps
};

export const useExamsData = () => {
  const [exampleValue, setExampleValue] = state.exampleValue(0);
  const examsList = useSelector(selectors.courseExamsList);
  const currentExam = useSelector(selectors.currentExam);
  const isLoading = reduxHooks.useRequestIsPending(RequestKeys.fetchCourseExams);

  return {
    currentExam,
    examsList,
    isLoading,
    exampleValue,
    setExampleValue,
  };
};

export const useExamAttemptsData = () => {
  const attemptsList = useSelector(selectors.courseExamAttemptsList);
  return { attemptsList };
};
