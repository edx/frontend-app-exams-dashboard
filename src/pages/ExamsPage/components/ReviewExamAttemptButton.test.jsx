import { render, screen } from '@testing-library/react';

import * as constants from 'data/constants';
import ReviewExamAttemptButton from './ReviewExamAttemptButton';

import * as hooks from '../hooks';

jest.mock('../hooks', () => ({
  useModifyExamAttempt: jest.fn(),
}));

const mockMakeNetworkRequest = jest.fn();

// nomally mocked for unit tests but required for rendering/snapshots
jest.unmock('react');

const reviewButton = (
  <ReviewExamAttemptButton
    username="username"
    examName="examName"
    attemptId={0}
    severity={1.0}
    submissionReason="Submitted by user"
  />
);

describe('ReviewExamAttemptButton', () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    hooks.useModifyExamAttempt.mockReturnValue(mockMakeNetworkRequest);
  });
  it('Test that the ReviewExamAttemptButton matches snapshot', () => {
    expect(render(reviewButton)).toMatchSnapshot();
  });
  it('Modal appears upon clicking button', () => {
    render(reviewButton);
    screen.getByText('Second Review Required').click();
    expect(screen.getByText('Please confirm your choice.')).toBeInTheDocument();
  });
  it('Clicking the No button closes the modal', () => {
    render(reviewButton);
    screen.getByText('Second Review Required').click();
    screen.getByText('No (Cancel)').click();
    // Using queryByText here allows the function to throw
    expect(screen.queryByText('Please confirm your choice.')).not.toBeInTheDocument();
  });
  it('Clicking the Verify button calls the modify exam attempt hook', () => {
    const mockModifyExamAttempt = jest.fn();
    jest.spyOn(hooks, 'useModifyExamAttempt').mockImplementation(() => mockModifyExamAttempt);
    render(reviewButton);
    screen.getByText('Second Review Required').click();
    screen.getByText('Verify').click();
    expect(mockModifyExamAttempt).toHaveBeenCalledWith(0, constants.ExamAttemptActions.verify);
  });
  it('Clicking the Reject button calls the modify exam attempt hook', () => {
    const mockModifyExamAttempt = jest.fn();
    jest.spyOn(hooks, 'useModifyExamAttempt').mockImplementation(() => mockModifyExamAttempt);
    render(reviewButton);
    screen.getByText('Second Review Required').click();
    screen.getByText('Reject').click();
    expect(mockModifyExamAttempt).toHaveBeenCalledWith(0, constants.ExamAttemptActions.reject);
  });
});
