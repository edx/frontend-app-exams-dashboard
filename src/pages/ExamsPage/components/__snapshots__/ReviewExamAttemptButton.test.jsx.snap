// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewExamAttemptButton Test that the ReviewExamAttemptButton matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="d-flex text-danger"
      >
        <button
          class="text-danger btn btn-link btn-sm"
          type="button"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
              fill="currentColor"
            />
          </svg>
          Second Review Required
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="d-flex text-danger"
    >
      <button
        class="text-danger btn btn-link btn-sm"
        type="button"
      >
        <svg
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
            fill="currentColor"
          />
        </svg>
        Second Review Required
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
